processing/src/main/java/io/druid/query/aggregation/cardinality/CardinalityBufferAggregator.java

```diff
@@ -53,16 +53,23 @@ public void init(ByteBuffer buf, int position)
   @Override
   public void aggregate(ByteBuffer buf, int position)
   {
-    final HyperLogLogCollector collector = HyperLogLogCollector.makeCollector(
-        (ByteBuffer) buf.duplicate().position(position).limit(
-            position
-            + HyperLogLogCollector.getLatestNumBytesForDenseStorage()
-        )
-    );
-    if (byRow) {
-      CardinalityAggregator.hashRow(selectorList, collector);
-    } else {
-      CardinalityAggregator.hashValues(selectorList, collector);
+    // Save position, limit and restore later instead of allocating a new ByteBuffer
```

Comment:
I didn't test Cardinality, but I did test HyperUnique, and the performance boost from avoiding the `duplicate()` of the left-hand side is substantial (a significant chunk of the overall ~19% boost).